azavea/echo-locator
# Feature/kak/profile email registration#286

## Overview

Make the client email address a profile field and handle client account creation on profile save. Also stores client account state on profile: whether or not a client account exists for the email, and whether on not that client has logged in for the first time.


### Demo

Counselors can edit the email on a profile until a Cognito account has been successfully created for the email. After that, the email cannot be edited.

Once a counselor has saved a profile with a new email, a Cognito account gets created, but is not confirmed until the first time the client logs in. Until the first client login, the counselor can resend the invite from the profile:

![image](https://user-images.githubusercontent.com/960264/70641176-30f80500-1c0b-11ea-9fa4-8a8fc032f3aa.png)

After the client has logged in for the first time, confirming their Cognito account, counselors can no longer resend the sign-up invitation email with temporary password:
![image](https://user-images.githubusercontent.com/960264/70641299-6d2b6580-1c0b-11ea-9f59-eb812dd1f7b7.png)

Clients get a read-only view of their profile email. If a client deletes their profile, the site will log them out, then the next time the client logs in, create a new profile with defaults with the client's voucher number and email address pre-set:
![image](https://user-images.githubusercontent.com/960264/70641686-fe9ad780-1c0b-11ea-9ad6-1f776c601b1d.png)


If a counselor tries to create a profile with an email already associated with a Cognito account, but for a different voucher number than that of the new profile, a specific save error message tells the client what voucher number is assigned to the Cognito account for that email:

![image](https://user-images.githubusercontent.com/960264/70640957-db235d00-1c0a-11ea-9269-d91d08fe1988.png)

If a counselor creates a profile with an email already associated with a Congito account that has the same voucher number as the new profile, the counselor will be able to successfully save the new profile, and viewing the profile again will reflect whether the client has confirmed their account yet. As new profiles can only be created after searching for the voucher number, this should prevent counselors accidentally overwriting an existing profile.


### Notes

Counselors have no way to delete client Cognito accounts once they have been created, and client Cognito accounts are permanently associated with the voucher number assigned at the time the account was created, so if a counselor mistakenly creates a profile associating an email address with the wrong voucher number, there's no way to fix it. Adding a Cognito account deletion feature would address the issue, but would be non-trivial to implement.


## Testing Instructions

 * Log in as a counselor
 * Test creating, modifying, and deleting user profiles
 * Email field should only be editable until client account has been made
 * Counselors should be able to resend invites only until client first logs in
 * Appropriate error should show on attempting to create a profile with an email already associated with a different profile, whether or not the client with that email has logged in for the first time yet
 * Log in as a client
 * Counselor-created profile for client should be found if it exists
 * Email should always be set and be read-only for the client
 * If client does not have a profile (as when deleting their own), a new one should be created on next client log-in with the email and voucher number preset


Closes #286
